tuned/plugins/plugin_irq.py

```diff
@@ -0,0 +1,269 @@
+		super(IrqPlugin, self).__init__(monitor_repository, storage_factory, hardware_inventory, device_matcher, device_matcher_udev, plugin_instance_factory, global_cfg, variables)
+		self._irqs = {}
+
+	#
+	# plugin-level methods: devices and plugin options
+	#
+	def _init_devices(self):
+		"""Read /proc/irq to collect devices
+		"""
+		self._devices_supported = True
+		self._free_devices = set()
+		self._assigned_devices = set()
+		for i in os.listdir("/proc/irq"):
+			p = os.path.join("/proc/irq", i)
+			if os.path.isdir(p) and i.isdigit():
+				info = IrqInfo(i)
+				self._irqs[i] = info
+				self._free_devices.add(info.device)
+		default_info = IrqInfo("DEFAULT")
+		default_info.device = "DEFAULT"
+		self._irqs["DEFAULT"] = default_info
+		self._free_devices.add("DEFAULT")
+
+	def _get_device_instance(self, device):
+		for name, instance in self._instances.items():
+			if device in instance.assigned_devices | instance.processed_devices:
+				return instance
+		log.error("Could not find instance for device %s" % device)
+		return None
+
+	@classmethod
+	def _get_config_options(cls):
+		return {
+			"affinity": "",
+			"mode": "set",
+		}
+
+	#
+	# instance-level methods: implement the Instance interface
+	#
+	def _instance_init(self, instance):
+		instance._has_static_tuning = True
+		instance._has_dynamic_tuning = False
+
+		affinity = instance.options.get("affinity")
+		instance._affinity = set(self._cmd.cpulist_unpack(affinity))
+		if len(instance._affinity) == 0:
+			log.info("Instance '%s' has no affinity. Deactivating." % instance.name)
+			instance._active = False
+
+		mode = instance.options.get("mode")
+		if mode in ["set", "intersect"]:
+			instance._mode = mode
+		else:
+			log.warning("Invalid operating mode '%s' for instance %s. Using the default 'set' instead."
+						% (mode, instance.name))
+			instance._mode = "set"
+
+	def _instance_cleanup(self, instance):
+		pass
+
+	def _instance_apply_static(self, instance):
+		log.debug("Applying IRQ affinities (%s)" % instance.name)
+		super(IrqPlugin, self)._instance_apply_static(instance)
+
+	def _instance_unapply_static(self, instance, rollback):
+		log.debug("Unapplying IRQ affinities (%s)" % instance.name)
+		super(IrqPlugin, self)._instance_unapply_static(instance, rollback)
+
+	def _instance_verify_static(self, instance, ignore_missing, devices):
+		log.debug("Verifying IRQ affinities (%s)" % instance.name)
+		return super(IrqPlugin, self)._instance_verify_static(instance, ignore_missing, devices)
+
+	#
+	# "low-level" methods to get/set irq affinities
+	#
+	def _get_irq_affinity(self, irq):
+		"""Get current IRQ affinity from the kernel
+
+		Args:
+			irq (str): IRQ number (as string) or "DEFAULT"
+
+		Returns:
+			affinity (set): set of all CPUs that belong to the IRQ affinity mask
+		"""
+		filename = "/proc/irq/default_smp_affinity" if irq == "DEFAULT" else "/proc/irq/%s/smp_affinity" % irq
+		with open(filename, "r") as f:
```

Comment:
Yes. I added a try/catch, and decided to output a log.debug in case of failure. As we're reading only in connection with a write (to get the original value), and we can assume that the write will fail for the same reason, the user will still see the "Failed to set SMP affinity" error.